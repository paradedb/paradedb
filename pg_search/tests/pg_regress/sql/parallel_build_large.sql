\i common/parallel_build_large_setup.sql

SET max_parallel_workers = 8;
SET client_min_messages TO INFO;

DO $$
DECLARE
    maintenance_work_mem text[] := ARRAY['2GB', '64MB'];
    maintenance_workers int[] := ARRAY[16, 8, 2];
    leader_participation boolean[] := ARRAY[true, false];
    target_segments int[] := ARRAY[4, 32];
    mw int;
    lp boolean;
    ts int;
    mwm text;
    count_val int;
    num_docs_val int;
BEGIN
    FOREACH mw IN ARRAY maintenance_workers LOOP
        FOREACH lp IN ARRAY leader_participation LOOP
            FOREACH ts IN ARRAY target_segments LOOP
                FOREACH mwm IN ARRAY maintenance_work_mem LOOP
                    -- Set configuration
                    EXECUTE format('SET max_parallel_maintenance_workers = %s', mw);
                    EXECUTE format('SET parallel_leader_participation = %s', lp);
                    EXECUTE format('SET paradedb.target_segment_count = %s', ts);
                    EXECUTE format('SET maintenance_work_mem = %L', mwm);

                    CREATE INDEX parallel_build_large_idx ON parallel_build_large USING bm25 (id, name) WITH (key_field = 'id');

                    -- Check index info and display results
                    SELECT COUNT(*) INTO count_val FROM paradedb.index_info('parallel_build_large_idx');
                    SELECT SUM(num_docs) INTO num_docs_val FROM paradedb.index_info('parallel_build_large_idx');
                    RAISE INFO 'Config: workers=%, work_mem=%,leader_participation=%, segments=%-> Count: %, Num Docs: %',
                        mw, mwm,lp, ts, count_val, num_docs_val;

                    DROP INDEX parallel_build_large_idx;
                END LOOP;
            END LOOP;
        END LOOP;
    END LOOP;
END $$;


\i common/parallel_build_large_cleanup.sql
