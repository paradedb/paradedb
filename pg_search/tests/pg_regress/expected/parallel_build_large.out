\i common/parallel_build_large_setup.sql
DROP TABLE IF EXISTS parallel_build_large;
CREATE EXTENSION IF NOT EXISTS pg_search;
CREATE TABLE parallel_build_large (
    id SERIAL PRIMARY KEY,
    name TEXT
);
INSERT INTO parallel_build_large (name)
SELECT 'lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'
FROM generate_series(1, 35000);
SET max_parallel_workers = 8;
SET client_min_messages TO INFO;
-- This should return a "not enough memory" error
SET maintenance_work_mem = '64MB';
SET max_parallel_maintenance_workers = 8;
SET paradedb.target_segment_count = 16;
CREATE INDEX parallel_build_large_idx ON parallel_build_large USING bm25 (id, name) WITH (key_field = 'id');
ERROR:  `maintenance_work_mem` is not high enough to give each parallel worker 15MB
-- These should complete and create the target segment count
DO $$
DECLARE
    maintenance_work_mem text[] := ARRAY['2GB', '128MB'];
    maintenance_workers int[] := ARRAY[6, 2];
    leader_participation boolean[] := ARRAY[true, false];
    target_segments int[] := ARRAY[4, 32];
    mw int;
    lp boolean;
    ts int;
    mwm text;
    count_val int;
    num_docs_val int;
BEGIN
    FOREACH mw IN ARRAY maintenance_workers LOOP
        FOREACH lp IN ARRAY leader_participation LOOP
            FOREACH ts IN ARRAY target_segments LOOP
                FOREACH mwm IN ARRAY maintenance_work_mem LOOP
                    -- Set configuration
                    EXECUTE format('SET max_parallel_maintenance_workers = %s', mw);
                    EXECUTE format('SET parallel_leader_participation = %s', lp);
                    EXECUTE format('SET paradedb.target_segment_count = %s', ts);
                    EXECUTE format('SET maintenance_work_mem = %L', mwm);

                    CREATE INDEX parallel_build_large_idx ON parallel_build_large USING bm25 (id, name) WITH (key_field = 'id');

                    -- Check index info and display results
                    SELECT COUNT(*) INTO count_val FROM paradedb.index_info('parallel_build_large_idx');
                    SELECT SUM(num_docs) INTO num_docs_val FROM paradedb.index_info('parallel_build_large_idx');
                    RAISE INFO 'Config: workers=%, work_mem=%,leader_participation=%, segments=%-> Count: %, Num Docs: %',
                        mw, mwm,lp, ts, count_val, num_docs_val;

                    DROP INDEX parallel_build_large_idx;
                END LOOP;
            END LOOP;
        END LOOP;
    END LOOP;
END $$;
INFO:  Config: workers=6, work_mem=2GB,leader_participation=t, segments=4-> Count: 4, Num Docs: 35000
INFO:  Config: workers=6, work_mem=128MB,leader_participation=t, segments=4-> Count: 4, Num Docs: 35000
INFO:  Config: workers=6, work_mem=2GB,leader_participation=t, segments=32-> Count: 32, Num Docs: 35000
INFO:  Config: workers=6, work_mem=128MB,leader_participation=t, segments=32-> Count: 32, Num Docs: 35000
INFO:  Config: workers=6, work_mem=2GB,leader_participation=f, segments=4-> Count: 4, Num Docs: 35000
INFO:  Config: workers=6, work_mem=128MB,leader_participation=f, segments=4-> Count: 4, Num Docs: 35000
INFO:  Config: workers=6, work_mem=2GB,leader_participation=f, segments=32-> Count: 32, Num Docs: 35000
INFO:  Config: workers=6, work_mem=128MB,leader_participation=f, segments=32-> Count: 32, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=2GB,leader_participation=t, segments=4-> Count: 4, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=128MB,leader_participation=t, segments=4-> Count: 4, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=2GB,leader_participation=t, segments=32-> Count: 32, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=128MB,leader_participation=t, segments=32-> Count: 32, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=2GB,leader_participation=f, segments=4-> Count: 4, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=128MB,leader_participation=f, segments=4-> Count: 4, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=2GB,leader_participation=f, segments=32-> Count: 32, Num Docs: 35000
WARNING:  only 2 parallel workers were available for index build
INFO:  Config: workers=2, work_mem=128MB,leader_participation=f, segments=32-> Count: 32, Num Docs: 35000
\i common/parallel_build_large_cleanup.sql
DROP TABLE IF EXISTS parallel_build_large;
